src/server/async_disconnect_handler.cpp

```diff
@@ -92,11 +92,12 @@ async_disconnect_handler<T_CONN_HANDLER>::disconnect_loop ()
 	m_disconnect_queue.swap (disconnect_work_buffer);
       }
 
-      while (!disconnect_work_buffer.empty ())
-	{
-	  connection_handler_uptr_t &front = disconnect_work_buffer.front ();
-	  front.reset (nullptr);
-	  disconnect_work_buffer.pop ();
-	}
+      disconnect_work_buffer = {}; // clear
+    }
+
+  // clear requests added after swapped to m_disconnect_queue before termination.
+  if (!m_disconnect_queue.empty ())
```

Comment:
maybe just add a comment that the queue can be accessed here unlocked as `disconnect` cannot be accessed by other thread because `m_terminate` is already set.

Reply:
Actually, it can. Because during 95L`disconnect_work_buffer = {};`, the lock is released, so `disconnect()` can acquire the lock and interleave in. When it's about to push the disconnect request, This clearing `m_disconnect_queue` job can be tried. This also has to be guarded..

Reply:
I've made the change to also acquire the lock during clearing left requests.

Reply:
I think that the problem is the atomic `m_terminate`. The code would be more free of inconsistent states if the `m_terminate` is a simple bool that will be guarded by the same mutex as the container. And `disconnect_loop` function reworked a little.
But since it is not super-critical code - ie: it is used occasionally when a xTS will disconnect, let's leave it like this.